### DISCUSSION TITLE
Avoiding as.rsi() changing D to I automatically

### DISCUSSION THREAD

**Discussion by send2dan** [msberends/AMR#80] [Q&A] [answered]
Hello,

RE: 

`as.rsi(c("S",` "I", "R", "D", "U", "C"))`

as.rsI() appears to convert D and U to I, which I don't actually want - Is there a way to switch this off, so that 

`as.rsi(c("S", "I", "R", "D", "U", "C"))`

leads to
Class <rsi>
[1] S    I    R    <NA> <NA> <NA>

rather than
Class <rsi>
[1] S    I    R    I    S    <NA>

Even better, is there any way that D can be retained as D? e.g.
Class <rsi>
[1] S    I    R    D <NA> <NA>

**Comment by msberends** [comment]
Not at the moment. The reason why it’s made I is here:

https://github.com/msberends/AMR/blob/05d536ebb7e1897cf0606c1308ff9b3bbdf0239f/R/rsi.R#L350-L351

Since EUCAST uses I as the dose-dependent interpretation since 2019, it’s considered I. 

What would be reasons to show D as well? We could add it of course, but then I would be curious for the reasons why.

**Comment by send2dan** [comment]
Thanks for your very helpful reply!

Good question - It's for purely local reasons, so I wouldn't suggest changing your code. We are using a very old LIMS here in Newcastle, and we are limited with regards to the functionality of our system, including the values entered. Therefore, for purely historical reasons (which are not worth delving into), we have on occasion used "D" to signify something other than dose-dependent. Thankfully, we are upgrading our LIMS over the next two years or so, to Cerner PathNet/LIMS. 

Best wishes,

Dan

**Comment by msberends** [comment, marked as answer]
Ah, tell me about it! Old lab systems are a pain!

But right, then it would be quite hard to work with D in the AMR package.

I even tried doing this for you:

```r
my_own_rsi <- function(x, include_these = c("S", "I", "D", "R")) {
  structure(
    factor(x, levels = include_these, ordered = TRUE),
    class = c("rsi", "factor")
  )
}
```

And then tried whether it could be used to determine resistance, but it doesn't:

```r
x <- my_own_rsi(c(rep("S", 100), rep("D", 100), rep("R", 100)))
# this seems to work:
x
#> Class 'rsi'
#>   [1] S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S S
#>  [73] S S S S S S S S S S S S S S S S S S S S S S S S S S S S D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D
#> [145] D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D D R R R R R R R R R R R R R R R R
#> [217] R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R R
#> [289] R R R R R R R R R R R R

# but then this doesn't - it's completely ignoring D:
resistance(x)
#> [1] 0.5
```

Maybe something like this is helpful then?

```r
data.frame(amox = x,
           cipr = x,
           gent = x,
           tobr = x) %>%
  summarise(across(aminoglycosides(),
                   function(x) sum(x == "D") / length(x)))
#> ℹ For aminoglycosides() using columns 'gent' (gentamicin) and 'tobr' (tobramycin)
#>        gent      tobr
#> 1 0.3333333 0.3333333
```

It calculates the proportion of D, but using [antibiotic selectors](https://msberends.github.io/AMR/reference/antibiotic_class_selectors.html) so you can still use some of the AMR functions.

Or for the proportion S + I + D with a manual function:

```r
my_susceptibility <- function(x) {
  x <- x[!is.na(x)]
  if (length(x) < 30) {
    stop("You might not want to know the result with so few isolates!")
  }
  sum(x %in% c("S", "I", "D")) / length(x)
}
data.frame(amox = x,
           cipr = x,
           gent = x,
           tobr = x) %>%
  summarise(across(c(quinolones(), aminoglycosides()),
                   my_susceptibility))
#> ℹ For quinolones() using column 'cipr' (ciprofloxacin)
#> ℹ For aminoglycosides() using columns 'gent' (gentamicin) and 'tobr' (tobramycin)
#>        cipr      gent      tobr
#> 1 0.6666667 0.6666667 0.6666667
```
